pkgs/build-support/node/fetch-pnpm-deps/default.nix

```diff
@@ -0,0 +1,87 @@
+{ lib
+, stdenvNoCC
+, nodePackages
+, jq
+, moreutils
+, cacert
+, makeSetupHook
+}: {
+  fetchPnpmDeps =
+    { src
+    , hash ? ""
+    , pname
+    , pnpm ? nodePackages.pnpm
+    , supportedArchitectures ? {
+        # We have to specify all possible platforms here, as pnpm doesn't allow us to specify a wildcard here
+        # https://github.com/pnpm/pnpm/issues/7695
+        # TODO: figure out remaining keys here
+        os = [ "win32" "darwin" "linux" ];
+        cpu = [ "x64" "arm64" ];
+      }
+    , ...
+    } @ args:
+    let
+      args' = builtins.removeAttrs args [ "hash" "pname" "pnpm" "supportedArchitectures" ];
+      hash' =
+        if hash != "" then {
+          outputHash = hash;
+        } else {
+          outputHash = "";
+          outputHashAlgo = "sha256";
+        };
+
+      # To be patched into package.json to make pnpm download optional platform dependencies
+      pnpmPatch = builtins.toJSON { pnpm = { inherit supportedArchitectures; }; };
+    in
+    # NOTE: This requires pnpm 8.10.0 or newer
+      # https://github.com/pnpm/pnpm/pull/7214
+    assert lib.versionAtLeast pnpm.version "8.10.0";
+    stdenvNoCC.mkDerivation (args' // {
+      name = "${pname}-pnpm-deps";
+
+      nativeBuildInputs = [
+        jq
+        moreutils
+        pnpm
+        cacert
+      ];
+
+      postPatch = ''
+        mv package.json package.json.orig
+        jq --raw-output '. * ${pnpmPatch}' package.json.orig > package.json
+      '';
+
+      # https://github.com/NixOS/nixpkgs/blob/763e59ffedb5c25774387bf99bc725df5df82d10/pkgs/applications/misc/pot/default.nix#L56
+      installPhase = ''
+        runHook preInstall
+
+        export HOME=$(mktemp -d)
+        pnpm config set store-dir $out
+        pnpm install --frozen-lockfile --ignore-script
```

Comment:
What guarantee do we have that this is reproducible across versions of pnpm?

Reply:
I don't think we have any guarantees here. So far `vesktop` has been using this, and it seems to be fairly stable

Reply:
I don't think this should prevent the PR from merging. There is no guarantee between versions of Maven either, but we can always update the hashes in `buildMavenPackage` treewide in the PR that updates `maven`.

Reply:
I have now pinned the pnpm version used by this function. This will almost certainly guarantee reasonable<sup>1</sup> reproducibility.

<sup>1</sup>: Perhaps different Node.js versions could influence the reproducibility, but I would argue that it's reasonably reproducible nevertheless.